<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>ETC3250: High Dimensional DA</title>
    <meta charset="utf-8" />
    <meta name="author" content="  Professor Di Cook     Econometrics and Business Statistics   Monash University" />
    <link href="libs/remark-css/kunoichi.css" rel="stylesheet" />
    <link href="libs/remark-css/ninjutsu.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link href="libs/countdown/countdown.css" rel="stylesheet" />
    <script src="libs/countdown/countdown.js"></script>
    <link rel="stylesheet" href="mystyle.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# ETC3250: High Dimensional DA
## Semester 1, 2020
### <br> Professor Di Cook <br> <br> Econometrics and Business Statistics <br> Monash University
### Week 3 (a)

---




## High Dimensional Data

.orange[High dimensional data] is data that has *many more* variables than observations, that is, `\(p \gg n\)`. It occurs commonly in bioinformatics, when genetic studies often have many more information on genes than patients.

&lt;center&gt;
&lt;img src="images/microarray.png" style="width: 70%; align: center"/&gt;
&lt;/center&gt;
---
class: split-two

.column[.pad50px[

## SRBCT cancer prediction 

- The SRBCT dataset (Khan et al., 2001) looks at classifying 4 classes of different childhood tumours sharing similar visual features during routine histology.
- Data contains 83 microarray samples with 1586 features.
- .orange[**Goal**: to use DA techniques to classify cancer types based on ]


]]

.column[.content.vmiddle.center[

 &lt;img src="images/SRBCT-nature.jpg", width="70%"&gt;

.purple[Source:] [Nature](https://www.nature.com/articles/modpathol2016119)

]]

---
class: split-two

.column[.pad50px[

## Recall: Discriminant Analysis


- *Strictly* assumes the conditional distribution of the data, given class grouping, is .orange[multivariate normal].
- Available through `MASS` package in <i class="fab  fa-r-project "></i> with functions `lda` (common covariance) and `qda`. 

]]


.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-1-1.png" width="504" style="display: block; margin: auto;" /&gt;



]]

---

class: split-two

.column[.pad50px[

## Advantages of DA

<i class="fas  fa-check "></i> Intuitive, and easy to use.

<i class="fas  fa-check "></i> Describes data generating process as well as provide a classifier for new points.

.orange[but...]

]]
.column[.pad50px[

## Disadvantages of DA

 <i class="fas  fa-times "></i> Does not work when `\(p &gt; n\)` due to MLE covariance matrix estimates being singular.

]]
---


class: split-two

.column[.pad50px[

## Advantages of DA

<i class="fas  fa-check "></i> Intuitive, and easy to use.

<i class="fas  fa-check "></i> Describes data generating process as well as provide a classifier for new points.

.orange[but...]

]]
.column[.pad50px[

## Disadvantages of DA

 <i class="fas  fa-times "></i> Does not work when `\(p &gt; n\)` due to MLE covariance matrix estimates being singular.
 
 &lt;br&gt;

.green[So, what can we do for high dimensional data?]

]]


---

class: split-two

.column[.pad50px[

## Diagonal Discriminant Analysis

&lt;br&gt;
- The simplest form of regularisation assumes that the features are independent within each class. 
- Consider a *diagonal-covariance* LDA rule for classifying classes
- A special case of the naive-Bayes classifier

]]
.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-2-1.png" width="504" style="display: block; margin: auto;" /&gt;

]]


---
## Discriminant Function

It can be shown that the discriminant score for a new observation `\(\mathbf{x}^*\)` when the features are considered independent reduces to the following:

`$$\delta_k(\mathbf{x}^*) = - \sum_{j=1}^{p}\frac{(x_j^* - \bar{x}_{kj})^2}{s^2_j} + 2\log \pi_k.$$`

The classification rule is then

`$$C(\mathbf{x}^*) = \ell \quad \mbox{if} \quad \delta_{\ell}(\mathbf{x}^*) = \max_k \delta_k(\mathbf{x}^*).$$`


---

## DLDA in R

DLDA is easy to run, is found in the `dlda` function provided by the `sparsediscrim` package (hosted on GitHub)
---

## What features are driving prediction?

Often the focus of genomics studies is not only to provide predictions (whether it be for cancer type, or prognosis), but also to understand the underlying drivers of such predictions. In this case, we need to know what features are impotant in the prediction mechanism.

As such, a drawback of diagonal LDA (and QDA) is that it uses all of the features, and is not convinent for interpretation. 



---

## Filter features for prediction

To motivate the upcoming method, consider a binary classfication DLDA problem. 

One way we could establish which of the features are driving prediction would be to perform a two-sample `\(t\)`-test 

`$$t_{j} = \frac{\bar{x}_{1j} - \bar{x}_{0j}}{s_j}$$`
with the `\(t\)` statistic providing a measure of how significant the difference in class means for predictor `\(j\)`. 



---

## Filter features for prediction

.green[Think about it:] Using the `\(t\)` statistic -  `\(t_{j} = \frac{\bar{x}_{1j} - \bar{x}_{0j}}{s_j}\)` for all features, what is one way we can determine important features for prediction?


<div class="countdown" id="timer_5e089c4d" style="right:0;bottom:0;" data-warnwhen="0">
<code class="countdown-time"><span class="countdown-digits minutes">01</span><span class="countdown-digits colon">:</span><span class="countdown-digits seconds">00</span></code>
</div>
---

## Filter features for prediction

.orange[Answer:] Can consider filtering for features with `\(\lvert t_j \lvert &gt; 2\)`, as this is deemed significant at the 5% level.

&lt;img src="multiDA_files/figure-html/unnamed-chunk-4-1.png" width="504" style="display: block; margin: auto;" /&gt;

.font_tiny[Note - further consideration can be given to the issue of [*Multiple Testing*](https://en.wikipedia.org/wiki/Multiple_comparisons_problem)]

---

## Nearest Shrunken Centroids (NSC)

Now consider the following statistic,

&lt;br&gt;

`$$d_{kj} = \frac{\bar{x}_{kj} - \bar{x}_j}{m_k(s_j + s_0)} \quad \mbox{with} \quad m_k^2 = \frac{1}{N_k} - \frac{1}{N}$$`
and `\(s_0\)` a small value to protect `\(d_{kj}\)` from small expression values.

&lt;br&gt;
This statistic is a  measure for how significant the difference between the class `\(k\)` mean for predictor `\(j\)`, and the overall mean for predictor `\(j\)`.
---
## Nearest Shrunken Centroids (NSC)

&lt;br&gt;

&lt;br&gt;

.green.center[**Question:** How would we filter features using this statistic? What is an appropriate threshold?]

---

## Option 1 - Hard Thresholding

Suppose we threshold these values:

&lt;br&gt;

`$$d'_{kj} = d_{kj} \cdot I\{ \lvert d_{kj} \lvert  \geq \Delta \},$$`
where `\(\Delta\)` is a parameter chosen via cross-validation. 

&lt;br&gt;

Then `\(d'_{kj} = d_{kj}\)` if the standardised class mean is siginificantly different from the overall mean, and zero otherwise.

---

## Option 2 - Soft Thresholding

Each `\(d_{kj}\)` is reduced by an amount `\(\Delta\)` in absolute value, and is set to zero if its absolute value is less than zero.

`$$d'_{kj} = \mbox{sign}(d_{kj})( \lvert d_{kj} \lvert - \Delta)_{+},$$`
&lt;center&gt;
 &lt;img src="images/soft.JPG", width="40%"&gt;
&lt;/center&gt;
---
## Nearest Shrunken Centroids Classifier

The NSC uses either version of the statistic `\(d'_{kj}\)` to regularise by shinking the class means towards the overall mean for each predictor seperately as follows:

`$$\bar{x}'_{kj} = \bar{x}_j + m_k(s_j + s_0)d'_{kj}$$`

&lt;br&gt;

.green.center[Unless a predictor has a significant difference to the overall mean for at least one class, it is useless for classification.]

We then use the shrunken centroids `\(\bar{x}'_{kj}\)` in place of `\(\bar{x}_{kj}\)` in the DLDA discriminant function.

---

## NSC in R

NSC can be used through the `pamr` function in R. It requires cross-validation to select the parameter `\(\Delta\)` - this is captured in the function `pamr.adaptthresh`.


```r
 p &lt;- ncol(x_train)
 mydata &lt;-  list(x=t(x_train),y=as.factor(y), geneid=1:p)

 #Initial training
 train_res &lt;-  pamr.train(mydata)
 new.scales &lt;-  pamr.adaptthresh(train_res)
 
 # Retrain with new delta value
 train_res &lt;-  pamr.train(mydata, threshold.scale=new.scales)
 predicted_values &lt;- pamr.predict(train_res, t(x_test), threshold=new.scales)
```



---
class: middle center

## Is there a method we can use to avoid lengthy CV tuning?

---

class: middle center

## Is there a method we can use to avoid lengthy CV tuning?

## .green[Yes!]

---
## multiDA (DA via multiple hypothesis testing)
---

class: center middle

Define, estimate, predict

---

class: split-two

.column[.pad50px[

## What defines a discriminative feature?
 
  
&lt;br&gt;

Suppose we have 3 classes to model. If we assume the features are independent,  within each feature we can group them as:



.orange[**One group**] (NOT a discriminative feature)

 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-6-1.png" width="504" style="display: block; margin: auto;" /&gt;
]]
---

class: split-two

.column[.pad50px[

## What defines a discriminative feature?
 
  
&lt;br&gt;

Suppose we have 3 classes to model. If we assume the features are independent,  within each feature we can group them as:



.orange[**Two groups**] (Groups 2 and 3, against 1)

 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-7-1.png" width="504" style="display: block; margin: auto;" /&gt;
]]

---
class: split-two

.column[.pad50px[

## What defines a discriminative feature?
 
  
&lt;br&gt;

Suppose we have 3 classes to model. If we assume the features are independent,  within each feature we can group them as:



.orange[**Two groups**] (Groups 1 and 3, against 2)

 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-8-1.png" width="504" style="display: block; margin: auto;" /&gt;
]]

---

class: split-two

.column[.pad50px[

## What defines a discriminative feature?
 
  
&lt;br&gt;

Suppose we have 3 classes to model. If we assume the features are independent,  within each feature we can group them as:



.orange[**Two groups**] (Groups 1 and 2, against 3)

 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-9-1.png" width="504" style="display: block; margin: auto;" /&gt;
]]

---

class: split-two

.column[.pad50px[

## What defines a discriminative feature?
 
  
&lt;br&gt;

Suppose we have 3 classes to model. If we assume the features are independent,  within each feature we can group them as:



.orange[**Three groups**] (All groups are different)

 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-10-1.png" width="504" style="display: block; margin: auto;" /&gt;
]]

---

class: split-two

.column[.pad50px[

## What about K &gt; 3 classes?
 
.green[Question: How many ways can you partition a set of size K?]
  
The total number of partitions of a K-element set is the [Bell number](https://en.wikipedia.org/wiki/Bell_number) `\(B_K\)`. The first several Bell numbers are:
`\(B_0= 1\)`, `\(B_1= 1\)`, `\(B_2= 2\)`, `\(B_3= 5\)`, `\(B_4= 15\)`, `\(B_5= 52\)`, and `\(B_6= 203\)`.

.font_small[.orange[Right: The 52 partitions of a set with 5 elements]]

 ]]

.column[.content.vmiddle.center[

&lt;img src="images/bell-number.png", width="40%"&gt;

]]
---

## Estimation

For each feature:

1. Estimate key paramters for Gaussian curves (mean, variance) and prior class probabilities
2. Estimate the probability of the appropriate partitioning. We represent this by a latent variable `\(\gamma_{jm}\)`.

&lt;br&gt;


It can be shown that estimates for 1) are MLEs.

---

## Latent variable estimation

The estimate for the latent variable `\(\gamma_{jm}\)` is expressed as follows:

&lt;br&gt;

$$
`\begin{align*}
\hat{\gamma}_{jm} = \frac{\exp[\frac{1}{2} \lambda_{jm}(\mathbf{x}_j, \mathbf{y}) + \log(\rho_m/\rho_1)] }{\sum_{\ell = 1}^M \exp[\frac{1}{2} \lambda_{j\ell}(\mathbf{x}_j, \mathbf{y}) + \log(\rho_{\ell}/\rho_1)]}
\end{align*}`
$$


---

## Latent variable estimation

The estimate for the latent variable `\(\gamma_{jm}\)` is expressed as follows:

&lt;br&gt;

$$
`\begin{align*}
\hat{\gamma}_{jm} = \frac{\exp[\frac{1}{2} \color{orange}{\lambda_{jm}}(\mathbf{x}_j, \mathbf{y}) + \log(\rho_m/\rho_1)] }{\sum_{\ell = 1}^M \exp[\frac{1}{2} \color{orange}{\lambda_{j\ell}}(\mathbf{x}_j, \mathbf{y}) + \log(\rho_{\ell}/\rho_1)]}
\end{align*}`
$$

&lt;br&gt;

- `\(\color{orange}{\lambda_{jm}}\)` are log likelihood ratio test (LRT) statistics;
---

## LRT – compare to the null

For all 5 hypotheses, compare the likelihood to the null. Pick the "partition" that is the most likely.

&lt;center&gt;

&lt;img src="images/LRT.png", width="80%"&gt;

&lt;/center&gt;
---


## Latent variable estimation

The estimate for the latent variable `\(\gamma_{jm}\)` is expressed as follows:

&lt;br&gt;

$$
`\begin{align*}
\hat{\gamma}_{jm} = \frac{\exp[\frac{1}{2} \lambda_{jm}(\mathbf{x}_j, \mathbf{y}) + \color{green}{\log(\rho_m/\rho_1)}] }{\sum_{\ell = 1}^M \exp[\frac{1}{2} \lambda_{j\ell}(\mathbf{x}_j, \mathbf{y}) + \color{green}{\log(\rho_{\ell}/\rho_1)}]}
\end{align*}`
$$

&lt;br&gt;

- `\(\lambda_{jm}\)` are log likelihood ratio test statistics;
- and `\(\color{green}{\log(\rho_m/\rho_1)}\)` can carefully be chosen such that an appropriate penalty can be applied. .orange[This means we need stronger evidence to support more complicated hypotheses.]

---

## A penalised likelihood ratio test statistic

Two forms of penalisation can be considered:

- .green[The BIC] - useful when Positive Selection Rate is preferred to controlling False Discovery Rate (FDR). 
`$$\nu_m \log(n)$$`
- .green[The Extended BIC] - useful for high dimensional data, penalising additionally on the number of features `\(p\)`.
`$$\nu_m[\log(n) + 2\log(p)]$$`

.font_small[(Note - `\\(\nu_m = g_m - 1\\)` where `\\(g_m\\)` is the number of groupings considered in model `\\(m\\)`). ]
---

class: split-60

.column[.pad50px[

## multiDA in R
 
&lt;br&gt;


```r
library(multiDA)
res &lt;- multiDA(y = SRBCT$y, 
               X = SRBCT$X,
               penalty = "EBIC",
               equal.var = TRUE,
               set.options = "exhaustive")
```

We can then examine the class groupings using the `plot()` method for `multiDA`:


```r
plot(res, ranks= 1)
```


 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-13-1.png" width="504" style="display: block; margin: auto;" /&gt;



]]


---

class: split-60

.column[.pad50px[

## multiDA in R
 
&lt;br&gt;


```r
library(multiDA)
res &lt;- multiDA(y = SRBCT$y, 
               X = SRBCT$X,
               penalty = "EBIC",
               equal.var = TRUE,
               set.options = "exhaustive")
```

This feature is one that `multiDA` is not as confident in - with the most likely seperation only having an estimated probability of 0.57.


```r
plot(res, ranks= 199)
```


 ]]

.column[.content.vmiddle.center[

&lt;img src="multiDA_files/figure-html/unnamed-chunk-16-1.png" width="504" style="display: block; margin: auto;" /&gt;



]]

---

## Compare performance - 100 trial, 5 fold CV

&lt;img src="multiDA_files/figure-html/unnamed-chunk-17-1.png" width="576" style="display: block; margin: auto;" /&gt;

---
layout: false
# 👩‍💻 Made by a human with a computer

### Slides at [https://monba.dicook.org](https://monba.dicook.org).
### Code and data at [https://github.com/dicook/Business_Analytics](https://github.com/dicook/Business_Analytics).
&lt;br&gt;

### Created using [R Markdown](https://rmarkdown.rstudio.com) with flair by [**xaringan**](https://github.com/yihui/xaringan), and [**kunoichi** (female ninja) style](https://github.com/emitanaka/ninja-theme).

&lt;br&gt; 
&lt;a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"&gt;&lt;img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /&gt;&lt;/a&gt;&lt;br /&gt;This work is licensed under a &lt;a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"&gt;Creative Commons Attribution-ShareAlike 4.0 International License&lt;/a&gt;.
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
